File: src/uniproxy/clash/rules.py
from __future__ import annotations

from typing import Literal

from attrs import define

from .base import BaseBasicRule, BaseRule, ProtocolLike


@define(slots=False)
class _NoResoleMixin:
    no_resolve: bool | None = None


@define
class DomainRule(BaseBasicRule):
    type: Literal["domain"] = "domain"


@define
class DomainSuffixRule(BaseBasicRule):
    type: Literal["domain-suffix"] = "domain-suffix"


@define
class DomainKeywordRule(BaseBasicRule):
    type: Literal["domain-keyword"] = "domain-keyword"


@define
class DomainSetRule(BaseBasicRule):
    type: Literal["domain-set"] = "domain-set"


@define
class IPCidrRule(_NoResoleMixin, BaseBasicRule):
    type: Literal["ip-cidr"] = "ip-cidr"

    def __str__(self) -> str:
        if self.no_resolve:
            return f"{self.type.upper()},{self.matcher},{self.policy},no-resolve"
        else:
            return f"{self.type.upper()},{self.matcher},{self.policy}"


@define
class IPCidr6Rule(_NoResoleMixin, BaseBasicRule):
    type: Literal["ip-cidr6"] = "ip-cidr6"

    def __str__(self) -> str:
        if self.no_resolve:
            return f"{self.type.upper()},{self.matcher},{self.policy},no-resolve"
        else:
            return f"{self.type.upper()},{self.matcher},{self.policy}"


@define
class GeoIPRule(_NoResoleMixin, BaseBasicRule):
    type: Literal["geoip"] = "geoip"

    def __str__(self) -> str:
        if self.no_resolve:
            return f"{self.type.upper()},{self.matcher},{self.policy},no-resolve"
        else:
            return f"{self.type.upper()},{self.matcher},{self.policy}"


@define
class UserAgentRule(BaseBasicRule):
    type: Literal["user-agent"] = "user-agent"


@define
class UrlRegexRule(BaseBasicRule):
    type: Literal["url-regex"] = "url-regex"


@define
class ProcessNameRule(BaseBasicRule):
    type: Literal["process-name"] = "process-name"


@define
class AndRule(BaseBasicRule):
    type: Literal["and"] = "and"


@define
class OrRule(BaseBasicRule):
    type: Literal["or"] = "or"


@define
class NotRule(BaseBasicRule):
    type: Literal["not"] = "not"


@define
class SubnetRule(BaseBasicRule):
    type: Literal["subnet"] = "subnet"


@define
class DestPortRule(BaseBasicRule):
    type: Literal["dest-port"] = "dest-port"


@define
class SrcPortRule(BaseBasicRule):
    type: Literal["src-port"] = "src-port"


@define
class InPortRule(BaseBasicRule):
    type: Literal["in-port"] = "in-port"


@define
class SrcIPRule(BaseBasicRule):
    type: Literal["src-ip"] = "src-ip"


@define
class ProtocolRule(BaseBasicRule):
    type: Literal["protocol"] = "protocol"


@define
class ScriptRule(BaseBasicRule):
    type: Literal["script"] = "script"


@define
class CellularRadioRule(BaseBasicRule):
    type: Literal["cellular-radio"] = "cellular-radio"


@define
class DeviceNameRule(BaseBasicRule):
    type: Literal["device-name"] = "device-name"


@define
class RuleSetRule(BaseBasicRule):
    type: Literal["rule-set"] = "rule-set"


@define
class FinalRule(BaseRule):
    policy: ProtocolLike
    type: Literal["final"] = "final"

    def __str__(self) -> str:
        return f"MATCH,{self.policy}"


type _BasicRule = (
    DomainRule
    | DomainSuffixRule
    | DomainKeywordRule
    | IPCidrRule
    | IPCidr6Rule
    | GeoIPRule
    | UserAgentRule
    | UrlRegexRule
    | ProcessNameRule
    | AndRule
    | OrRule
    | NotRule
    | SubnetRule
    | DestPortRule
    | InPortRule
    | SrcPortRule
    | SrcIPRule
    | ProtocolRule
    | ScriptRule
    | CellularRadioRule
    | DeviceNameRule
    | RuleSetRule
    | DomainSetRule
)

type Rule = _BasicRule | FinalRule

File: uniproxy/src/rules.scala
// scalafmt: { maxColumn = 180, align.preset = more }
package uniproxy
package rules

import uniproxy.typing.{BasicRuleType, GroupRuleType, RuleType}
import uniproxy.abc.{AbstractRule, ProtocolLike, RuleProviderLike}

type Matcher = RuleProviderLike
type Policy  = ProtocolLike

sealed trait Matchable

sealed trait BasicMatchable extends Matchable:
  val matcher: Matcher
  val policy: Policy
sealed trait BasicResolvedMatchable extends Matchable:
  val matcher: Matcher
  val policy: Policy
  val noResolve: Boolean
sealed trait GroupMatchable extends Matchable:
  val matchers: Seq[Matcher]
  val policy: Policy
sealed trait GroupResolvedMatchable extends Matchable:
  val matchers: Seq[Matcher]
  val policy: Policy
  val noResolve: Boolean

enum BasicRule(`type`: BasicRuleType) extends AbstractRule {

  case DomainRule(matcher: Matcher, policy: Policy)        extends BasicRule(BasicRuleType.domain) with BasicMatchable
  case DomainSuffixRule(matcher: Matcher, policy: Policy)  extends BasicRule(BasicRuleType.`domain-suffix`) with BasicMatchable
  case DomainKeywordRule(matcher: Matcher, policy: Policy) extends BasicRule(BasicRuleType.`domain-keyword`) with BasicMatchable

  case IPCidrRule(matcher: Matcher, policy: Policy, noResolve: Boolean = true)  extends BasicRule(BasicRuleType.`ip-cidr`) with BasicResolvedMatchable
  case IPCidr6Rule(matcher: Matcher, policy: Policy, noResolve: Boolean = true) extends BasicRule(BasicRuleType.`ip-cidr6`) with BasicResolvedMatchable
  case IPAsnRule(matcher: Matcher, policy: Policy, noResolve: Boolean = true)   extends BasicRule(BasicRuleType.`ip-asn`) with BasicResolvedMatchable
  case GeoIPRule(matcher: Matcher, policy: Policy, noResolve: Boolean = true)   extends BasicRule(BasicRuleType.`geoip`) with BasicResolvedMatchable

  case UserAgentRule(matcher: Matcher, policy: Policy)   extends BasicRule(BasicRuleType.`user-agent`) with BasicMatchable
  case UrlRegexRule(matcher: Matcher, policy: Policy)    extends BasicRule(BasicRuleType.`url-regex`) with BasicMatchable
  case ProcessNameRule(matcher: Matcher, policy: Policy) extends BasicRule(BasicRuleType.`process-name`) with BasicMatchable

  case AndRule(matcher: Matcher, policy: Policy) extends BasicRule(BasicRuleType.and) with BasicMatchable
  case OrRule(matcher: Matcher, policy: Policy)  extends BasicRule(BasicRuleType.or) with BasicMatchable
  case NotRule(matcher: Matcher, policy: Policy) extends BasicRule(BasicRuleType.not) with BasicMatchable

  case SubnetRule(matcher: Matcher, policy: Policy)        extends BasicRule(BasicRuleType.subnet) with BasicMatchable
  case DestPortRule(matcher: Matcher, policy: Policy)      extends BasicRule(BasicRuleType.`dest-port`) with BasicMatchable
  case SrcPortRule(matcher: Matcher, policy: Policy)       extends BasicRule(BasicRuleType.`src-port`) with BasicMatchable
  case InPortRule(matcher: Matcher, policy: Policy)        extends BasicRule(BasicRuleType.`in-port`) with BasicMatchable
  case SrcIPRule(matcher: Matcher, policy: Policy)         extends BasicRule(BasicRuleType.`src-ip`) with BasicMatchable
  case ProtocolRule(matcher: Matcher, policy: Policy)      extends BasicRule(BasicRuleType.protocol) with BasicMatchable
  case ScriptRule(matcher: Matcher, policy: Policy)        extends BasicRule(BasicRuleType.script) with BasicMatchable
  case CellularRadioRule(matcher: Matcher, policy: Policy) extends BasicRule(BasicRuleType.`cellular-radio`) with BasicMatchable
  case DeviceNameRule(matcher: Matcher, policy: Policy)    extends BasicRule(BasicRuleType.`device-name`) with BasicMatchable

  case DomainSetRule(matcher: Matcher, policy: Policy) extends BasicRule(BasicRuleType.`domain-set`) with BasicMatchable
  case RuleSetRule(matcher: Matcher, policy: Policy)   extends BasicRule(BasicRuleType.`rule-set`) with BasicMatchable

  case FinalRule(policy: Policy, dnsFailed: Boolean = true) extends BasicRule(BasicRuleType.`final`)

  override def toString(): String = this match
    case r: BasicResolvedMatchable if r.noResolve      => s"${`type`.toString.toUpperCase},${r.matcher},${r.policy},no-resolve"
    case r: BasicResolvedMatchable if !r.noResolve     => s"${`type`.toString.toUpperCase},${r.matcher},${r.policy}"
    case r: BasicMatchable                             => s"${`type`.toString.toUpperCase},${r.matcher},${r.policy}"
    case uniproxy.rules.BasicRule.IPCidrRule(_, _, _)  => ???
    case uniproxy.rules.BasicRule.IPCidr6Rule(_, _, _) => ???
    case uniproxy.rules.BasicRule.IPAsnRule(_, _, _)   => ???
    case uniproxy.rules.BasicRule.GeoIPRule(_, _, _)   => ???
    case FinalRule(policy, dnsFailed)                  => s"${`type`.toString.toUpperCase},${policy}" ++ (if dnsFailed then ",dns-failed" else "")
}

enum GroupRule(`type`: GroupRuleType) extends AbstractRule {

  def toBasicRules(): Seq[BasicRule] = this match
    case DomainGroupRule(matchers, policy)             => matchers.map(DomainRule(_, policy))
    case DomainSuffixGroupRule(matchers, policy)       => matchers.map(DomainSuffixRule(_, policy))
    case DomainKeywordGroupRule(matchers, policy)      => matchers.map(DomainKeywordRule(_, policy))
    case IPCidrGroupRule(matchers, policy, noResolve)  => matchers.map(IPCidrRule(_, policy, noResolve))
    case IPCidr6GroupRule(matchers, policy, noResolve) => matchers.map(IPCidr6Rule(_, policy, noResolve))

  case DomainGroupRule(matchers: Seq[Matcher], policy: Policy)        extends GroupRule(GroupRuleType.`domain-group`) with GroupMatchable
  case DomainSuffixGroupRule(matchers: Seq[Matcher], policy: Policy)  extends GroupRule(GroupRuleType.`domain-suffix-group`) with GroupMatchable
  case DomainKeywordGroupRule(matchers: Seq[Matcher], policy: Policy) extends GroupRule(GroupRuleType.`domain-keyword-group`) with GroupMatchable

  case IPCidrGroupRule(matchers: Seq[Matcher], policy: Policy, noResolve: Boolean = true)  extends GroupRule(GroupRuleType.`ip-cidr-group`) with GroupResolvedMatchable
  case IPCidr6GroupRule(matchers: Seq[Matcher], policy: Policy, noResolve: Boolean = true) extends GroupRule(GroupRuleType.`ip-cidr6-group`) with GroupResolvedMatchable
}

type Rule = BasicRule | GroupRule

extension (rule: Rule)
  def `type`: RuleType = rule match
    case r: BasicRule => r.`type`
    case r: GroupRule => r.`type`

export BasicRule.*
export GroupRule.*
